File: 2872_maximum_number_of_k_divisible_components.cpp
class Solution {
public:
    int maxKDivisibleComponents(int n, vector<vector<int>>& edg, vector<int>& val, int k) {
        vector<vector<int>> g(n);
        vector<int> deg(n);
        if (n < 2) return 1;
        for (auto& e : edg) {
            g[e[0]].push_back(e[1]);
            g[e[1]].push_back(e[0]);
            deg[e[0]]++;
            deg[e[1]]++;
        }
        vector<long long> v(val.begin(), val.end());
        queue<int> q;
        for (int i = 0; i < n; i++) if (deg[i] == 1) q.push(i);
        int cnt = 0;
        while (!q.empty()) {
            int u = q.front();
            q.pop();
            deg[u]--;
            long long rem = 0;
            if (v[u] % k == 0) cnt++;
            else rem = v[u];
            for (int nbh : g[u]) {
                if (deg[nbh] == 0) continue;
                deg[nbh]--;
                v[nbh] += rem;
                if (deg[nbh] == 1) q.push(nbh);
            }
        }
        return cnt;
    }
};

File: 2872_maximum_number_of_k_divisible_components.java
class Solution {
    public int maxKDivisibleComponents(int n, int[][] edg, int[] val, int k) {
        List<List<Integer>> gr = new ArrayList<>();
        int[] deg = new int[n];
        if (n < 2) return 1;
        for (int i = 0; i < n; i++) gr.add(new ArrayList<>());
        for (int[] e : edg) {
            gr.get(e[0]).add(e[1]);
            gr.get(e[1]).add(e[0]);
            deg[e[0]]++;
            deg[e[1]]++;
        }
        long[] v = new long[n];
        for (int i = 0; i < n; i++) v[i] = (long) val[i]; 
        Queue<Integer> q = new LinkedList<>();
        for (int i = 0; i < n; i++) if (deg[i] == 1) q.offer(i);
        int cnt = 0;
        while (!q.isEmpty()) {
            int u = q.poll();
            deg[u]--;
            long tmp = 0;  
            if (v[u] % k == 0) cnt++;
            else tmp = v[u];
            for (int nb : gr.get(u)) {
                if (deg[nb] == 0) continue;
                deg[nb]--;
                v[nb] += tmp;
                if (deg[nb] == 1) q.offer(nb);
            }
        }
        return cnt;
    }
}
